### ISSUE TITLE
Smoothness of visualization with dynamic data (using DataSets)

### ISSUE THREAD

**Issue by vukk** [almende/vis#1164]
Hi,

I'm using vis to display logic solver optimization output on graph/network problems, each step is a new optimum found by the solver. At each step I make changes to nodes and edges DataSets with {queue:true} and flush.
In some problems the changes between two optimums are very large, and smoothness of the visualization in these cases is quite important to me. Otherwise the user has a hard time following what happened. Of course if two successive solutions have no commonalities then discrete jumps are unavoidable. I use both the forceAtlas2Based solver and hierarchical layout, depending on the problem at hand.

I've tuned the physics (timeStep, damping) so changes happen a bit slower on screen.

The forceAtlas2Based physics solver is quite good even in this case, but the network still experiences a "shock" when new nodes and edges are added (and perhaps old ones removed).
I think there are two issues causing this behaviour:
- the edges contracting quite violently immediately after they are added (and the nodes are far enough apart, which is often the case)
- the nodes being added to origin instead of a "smarter place" (also related to the previous issue)

For the first issue, I'm not sure if vis already gives a way to enhance the quality. It seems that springLength and springConstant as a fixed parameters won't help. Perhaps a separate spring contracting damping parameter? I'm grasping at straws here.
For the second issue the obvious solution is to find better initial positions for added nodes, but I don't know if there are any smart algorithms for that?
Since for my purposes having a few seconds of delay in between changes is not a problem, I was at some point thinking of running a separate physics engine, fixing the coordinates of current nodes, letting the new nodes find their place and copy their end coordinates when adding them for real to the network that is actually shown. But this seems very complicated to do.
Another slightly simpler idea was to calculate the average position of the nodes a new node would be connected to by edges, but what to do with connected nodes that are added at the same time and don't have coordinates.

For the hierarchical layout I have two specific problems:
- When a node changes level, the change is not smooth. The x or y coordinate is just set and the node jumps.
- Nodes can't hop over each other on the defined level, so a tree that could be laid out nicely will usually have overlapping edges when nodes and edges are added and removed. See here an example of the [final optimized solution of a certain phylogenetic tree](http://i.imgur.com/dABmHrj.png) and a [the same tree with nodes dragged manually](http://i.imgur.com/uDF3vit.png).

The hierarchical layout is fairly new so I understand that it's not yet polished for these use cases.
I tried to look for algorithms that modify the idea of forceAtlas/barnesHut to display trees, but did not find any.

In conclusion what I'm looking for:
- Ideas for solving the previous issues / enhancing the behaviour.
- Whether the solution could&should be contributed back or be one-off.

Related issues: #1139 (node positioning), #1120 (use for node animation between hierarchy levels?)

**Comment by AlexDM0** [comment]
Hi,

The contraction of the edges could be the bug reported in #1111. It has been solved in v4.6+. Which version are you using?

Vis does place the new node randomly near the origin. I do plan of implementing an algorithm for stabilization to get a good estimate of the new position but I don't have much time for it at the moment.

This could also be applied to positioning new nodes but we'll have to wait and see.

As for the hierarchical system there are a few things:
The nodes are initially positioned according to the levels. Then the physics solver takes over. The solver is tailored to the hierarchical view. There is a repulsion boundary between nodes so they CAN cross but won't because of the solver. The crossing however should be solved by now. If I generate some random networks here:
http://visjs.org/examples/network/layout/hierarchicalLayout.html

There is not a single crossing.

When adding  a new node, the entire tree has to be recalculated, reinitialized and stabilized. This gives a jittery look. The moving the node an an animation could be done theoretically but there are a number of issues there. For instance, we'd have to disable physics for the duration of the animation.

We'd have to calculate the positions of the nodes in the new stabilized position without showing the user, then use these positions to 'animate' towards. The current workflow is not designed for this but it could be an interesting addition. 

There are a lot of things that could be improved, and they will be in due time. I am open to any suggestions on algorithms that would improve any of this.

I would like to add the Kamada–Kawai algorithm for layout estimates for instance.

Regards,

Alex

**Comment by vukk** [comment]
> The contraction of the edges could be the bug reported in #1111. It has been solved in v4.6+. Which version are you using?

I was indeed using 4.5 so an older version. The behavior that is seen on the fiddle <http://jsbin.com/yojirupuwo/1/edit?html,output> at #1111 is similar to what I'm experiencing, but the difference is that I'm adding and removing edges on purpose in my visualization and thus some contraction is of course unavoidable for the network to settle into a new layout. I'd just like to lessen the shock.  
I've found that I can control the edge contraction a bit by forcing nodes to move slowly. Setting physics maxVelocity low and using the forceAtlas2Based solver makes things smoother, but I'd still like to control the edge springs a bit more, e.g. have a similar dampening setting as for node velocity. This is not as critical to me anymore though, as I'm now using coloring etc. to indicate changes too, so the user doesn't have to just follow the nodes by eye to see what happens/happened.
Removing maxVelocity and forceAtlas solver here <http://jsfiddle.net/vukk/uzpgheog/> will make the network jiggle a lot, as the edges initially contract very fast when they are added (contraction in the sense of physics or muscles, not graph theory).

> I do plan of implementing an algorithm for stabilization to get a good estimate of the new position but I don't have much time for it at the moment.

Do you have any ideas, e.g. on your notes, on how to implement such an algorithm?

> There is a repulsion boundary between nodes so they CAN cross but won't because of the solver.

Yep. In my use case I'd like to somehow change this. Essentially what's happening is this: <http://jsfiddle.net/vukk/jgt1v9vn/>.

> We'd have to calculate the positions of the nodes in the new stabilized position without showing the user, then use these positions to 'animate' towards. The current workflow is not designed for this but it could be an interesting addition.

This could be one option, but maybe something in the lines of modifying the force directed layout to output a tree would make sense for the current workflow. I tried looking if anyone has done this or has any tips for this and the only example I found was <http://bl.ocks.org/mbostock/1138500>. Though it would be nice to get the inner node and leaf levels essentially straight, which they are not in that example.

> I would like to add the Kamada–Kawai algorithm for layout estimates for instance.

Is there an advantage to using it instead of just running the force directed algorithm on the side somehow? It's faster perhaps? With a very quick glance it seems to be an iterative algorithm like force directed.

**Comment by AlexDM0** [comment]
Hi,

I have since implemented the kamada kawai algorithm and it is a great improvement in stabilisation time. Your issue however is with the hierarchical view, which is not influenced by this.

> Do you have any ideas, e.g. on your notes, on how to implement such an algorithm? 
this is the kamada kawai algorithm.

>Yep. In my use case I'd like to somehow change this. Essentially what's happening is this: http://jsfiddle.net/vukk/jgt1v9vn/.
I really don't think you want this, the repulsion boundary makes sure the spacing between the nodes is even. You'd need to reset the hierarchical sorting. This can be slow for large systems.
http://jsfiddle.net/jgt1v9vn/2/

>I tried looking if anyone has done this or has any tips for this and the only example I found was http://bl.ocks.org/mbostock/1138500. Though it would be nice to get the inner node and leaf levels essentially straight, which they are not in that example.
I'm not sure this would solve the problem, but we dont have any plans to look into this. The current system works for most usecases

>Is there an advantage to using it instead of just running the force directed algorithm on the side somehow? It's faster perhaps? With a very quick glance it seems to be an iterative algorithm like force directed.
yes, it is very fast for systems with less than 100 nodes, it cuts the amount of required iterations to stabilise in half. For larger systems, I cluster the system to reduce the amount of nodes, position and decluster to keep it fast.

Regards

**Comment by AlexDM0** [comment]
Hi,

Is this still an issue?

Regards
